perf/perf_nmem.py

```diff
@@ -0,0 +1,248 @@
+
+
+class perfNMEM(Test):
+    """
+    Testing perf nmem PMU
+    :avocado: tags=privileged,perf,pmu
+
+    1. Check for nmem PMUs in sysfs
+    2. Check for CONFIG_PAPR_SCM kernel configuration
+    3. Check for nmem PMU register messages in dmesg
+    4. Check PMUs found in sysfs and dmesg are matching
+    5. Count the number of PMU events
+    6. Run all events one by one per PMU
+    7. Run events by grouping per PMU
+    8. Run mixed PMU events - negative test case
+    9. Check PMU cpumask exists or not
+    10. Check cpumask is same for all the PMUs or not
+    11. Check cpumask by on and off cpu
+    """
+
+    def setUp(self):
+        # Install required packages
+        smm = SoftwareManager()
+        detected_distro = distro.detect()
+        if 'ppc64' not in detected_distro.arch:
+            self.cancel("Processor is not PowerPC")
+        deps = ['gcc', 'make']
+        if 'Ubuntu' in detected_distro.name:
+            deps.extend(['linux-tools-common', 'linux-tools-%s'
+                         % platform.uname()[2]])
+        elif detected_distro.name in ['rhel', 'SuSE', 'fedora', 'centos']:
+            deps.extend(['perf'])
+        else:
+            self.cancel("Install the package for perf supported by %s"
+                        % detected_distro.name)
+        for package in deps:
+            if not smm.check_installed(package) and not smm.install(package):
+                self.cancel('%s is needed for the test to be run' % package)
+
+        # Set required variables
+        self.base_dir = "/sys/devices/"
+        self.pmu_list = []
+        self.all_events = {}
+
+        # Collect the PMU names from sysfs dir.
+        for files in os.listdir(self.base_dir):
+            if "nmem" in files:
+                self.pmu_list.append(files)
+        if self.pmu_list:
+            self.log.info("Found PMUs: %s" % self.pmu_list)
+        else:
+            # If PMUs not found then cancel the test run
+            self.cancel("nmem PMUs not found")
+
+        # Once the PMUs are there, collect the list of events
+        output = process.system_output('perf list nmem', shell=True,
+                                       ignore_status=True)
+        for ln in output.decode().splitlines():
+            ln = ln.strip()
+            # Skip empty line and header
+            if not ln or "List of pre-defined events" in ln:
+                continue
+            else:
+                ln = ln.split('[')[0].strip()
+                for pmu in self.pmu_list:
+                    if ln.split('/')[0] in pmu:
+                        # Collecting the events per PMU
+                        # ex:{nmem0: [ev1, ev2, ...], nmem1: [ev1, ev2, ...]}
+                        self.all_events.setdefault(pmu, []).append(ln)
+        if not self.all_events:
+            self.cancel("Can not extract events from 'perf list nmem'")
+
+    def test_kernel_config(self):
+        cfg_param = "CONFIG_PAPR_SCM"
```

Comment:
one help is this is prerequisite of test or just a one test scenario ?

Reply:
Should be pre-requisite. changed the code.